Failed containers do not fail the build, but failed assumptions do

## Overview

The following logic is incorrect.

https://github.com/graalvm/native-build-tools/blob/15bea361faf74d2f36655aef688995be17e59918/common/junit-platform-native/src/main/java/org/graalvm/junit/platform/NativeImageJUnitLauncher.java#L137

The `failedCount` should include failed tests and failed containers (i.e., test classes), but it should not include aborted tests.

An "aborted test" is one that threw an `org.junit.AssumptionViolatedException` in JUnit 4 or an `org.opentest4j.TestAbortedException` in JUnit Jupiter (JUnit 5), typically via `org.junit.Assume` and `org.junit.jupiter.api.Assumptions` respectively, but such exceptions do not indicate that the test failed. Rather, they indicate that the test was aborted mid-flight -- for example, due to a constraint that cannot be fulfilled in the current environment.

The reason I noticed this is that I encountered the following in the build for Spring Native when running a sample application with the Maven plugin for Native Build Tools.

```
Failures (1):
  JUnit Jupiter:CommandlinerunnerApplicationTests
    ClassSource [className = 'com.example.commandlinerunner.CommandlinerunnerApplicationTests', filePosition = null]
    => org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.test.context.web.ServletTestExecutionListener]: No default constructor found; nested exception is java.lang.NoSuchMethodException: org.springframework.test.context.web.ServletTestExecutionListener.<init>()
       org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:29)
       org.springframework.test.context.support.AbstractTestContextBootstrapper.instantiateListeners(AbstractTestContextBootstrapper.java:183)
       org.springframework.test.context.support.AbstractTestContextBootstrapper.getTestExecutionListeners(AbstractTestContextBootstrapper.java:167)
       org.springframework.test.context.TestContextManager.<init>(TestContextManager.java:138)
       org.springframework.test.context.TestContextManager.<init>(TestContextManager.java:122)
       [...]
     Caused by: java.lang.NoSuchMethodException: org.springframework.test.context.web.ServletTestExecutionListener.<init>()
       java.lang.Class.getConstructor0(DynamicHub.java:3349)
       java.lang.Class.getDeclaredConstructor(DynamicHub.java:2553)
       org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:25)
       [...]

Test run finished after 34 ms
[         2 containers found      ]
[         0 containers skipped    ]
[         2 containers started    ]
[         0 containers aborted    ]
[         1 containers successful ]
[         1 containers failed     ]
[         1 tests found           ]
[         0 tests skipped         ]
[         0 tests started         ]
[         0 tests aborted         ]
[         0 tests successful      ]
[         0 tests failed          ]

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
```

The build was successful even though none of the tests executed (because the containing test class failed).

## Solution

Change:

```java
long failedCount = summary.getTestsFailedCount() + summary.getTestsAbortedCount();
```

To:

```java
long failedCount = summary.getTotalFailureCount();
```

As stated in the Javadoc for `getTotalFailureCount()`:

> Get the total number of failed containers and failed tests.

Note that the `ConsoleLauncher` in the JUnit Platform calculates the exit code similarly:

https://github.com/junit-team/junit5/blob/f4ef4b9e607579dd0000d4bfcfe6a34b39823b53/junit-platform-console/src/main/java/org/junit/platform/console/ConsoleLauncherExecutionResult.java#L47-L52

## Deliverables

- [x] aborted tests must **not** fail the build
- [x] failed containers **must** fail the build

--- a/common/junit-platform-native/src/main/java/org/graalvm/junit/platform/NativeImageJUnitLauncher.java
+++ b/common/junit-platform-native/src/main/java/org/graalvm/junit/platform/NativeImageJUnitLauncher.java
@@ -134,7 +134,7 @@ public class NativeImageJUnitLauncher {
             summary.printTo(out);
         }
 
-        long failedCount = summary.getTestsFailedCount() + summary.getTestsAbortedCount();
+        long failedCount = summary.getTotalFailureCount();
         System.exit(failedCount > 0 ? 1 : 0);
     }
 }

--- a/common/junit-platform-native/src/test/java/org/graalvm/junit/jupiter/BasicTests.java
+++ b/common/junit-platform-native/src/test/java/org/graalvm/junit/jupiter/BasicTests.java
@@ -42,6 +42,9 @@
 package org.graalvm.junit.jupiter;
 
 import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Assumptions;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Disabled;
 import org.junit.jupiter.api.DisplayName;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.TestInfo;
@@ -52,15 +55,36 @@ class BasicTests {
         System.out.println("Running test: " + info.getDisplayName());
     }
 
+    @BeforeAll
+    static void beforeAll() {
+        // Uncomment this block to see that a failed container fails the build.
+        //
+        // if (System.getProperty("org.graalvm.nativeimage.imagecode") != null) {
+        //     throw new RuntimeException("Failed test container within native image");
+        // }
+    }
+
+    @Test
+    @Disabled
+    void skippedTest() {
+        throw new RuntimeException("Test must have been disabled/skipped entirely");
+    }
+
+    @Test
+    void abortedTest() {
+        Assumptions.assumeTrue(false, "false is never true");
+        throw new RuntimeException("Test must have been aborted mid-flight");
+    }
+
     @Test
     @DisplayName("Basic test one")
-    void addsTwoNumbers(TestInfo testInfo) {
+    void basicTest1(TestInfo testInfo) {
         Assertions.assertEquals("Basic test one", testInfo.getDisplayName());
     }
 
     @Test
     @DisplayName("Basic test two")
-    void addsTwoNumbers2(TestInfo testInfo) {
+    void basicTest2(TestInfo testInfo) {
         Assertions.assertEquals("Basic test two", testInfo.getDisplayName());
     }
 

